File: mcr_robinson/common/scripts/mcr_robinson_python/ComponentContainer.py

from mcr_robinson_python.DataPort import DataPortOutput

class State:
    IDLE = 0
    ACTIVE = 1

class Mode:
    ON_UPDATE = 0
    CYCLIC = 1
        
class ComponentContainer:

    
    state = State.IDLE
    init = False
    mode = Mode.ON_UPDATE
    
    updateEvent = DataPortOutput();
    initEvent = DataPortOutput();
    shutdownEvent = DataPortOutput();
    
    def __init__(self, component):
        self.component = component
        
        self.updateEvent.connect(component.update)
        self.initEvent.connect(component.init)
        self.shutdownEvent.connect(component.shutdown)
    
    def __del(self):
        self.shutdownEvent()
        
    def eventport_input_start(self):
        print("ComponentContainer::eventport_input_start")
        
        self.state = State.ACTIVE
        
        if (self.init == False):
            self.initEvent()
            self.init = True
            
    def eventport_input_stop(self):
        print("ComponentContainer::eventport_input_stop")
        
        self.state = State.IDLE

    def update(self):
        if (self.state == State.ACTIVE):
            self.updateEvent()

    def update_on_data(self):
        if (self.mode == Mode.ON_UPDATE):
            self.update()


    def update_cycle(self):
        if (self.mode == Mode.CYCLIC):
            self.update()

File: mcr_robinson_java/robinson/src/main/java/mcr/robinson/ComponentContainer.java
package mcr.robinson;

import java.lang.reflect.Method;

import mcr.robinson.annotations.EventInputPort;
import mcr.robinson.annotations.EventOutputPortInfo;
import mcr.robinson.annotations.Init;
import mcr.robinson.annotations.Shutdown;
import mcr.robinson.annotations.Update;


public class ComponentContainer {

	public enum State {IDLE, ACTIVE}
	
	public enum Mode {CYCLIC, ON_EVENT}
	
	Object component;
	
	Class<?> interfaceClass;
	
	@EventOutputPortInfo(name="event_out", event="e_update")
	EventOutputPort event_update = new EventOutputPort();
	
	@EventOutputPortInfo(name="event_out", event="e_init")
	EventOutputPort event_init = new EventOutputPort();
	
	@EventOutputPortInfo(name="event_out", event="e_shutdown")
	EventOutputPort event_shutdown = new EventOutputPort();
	
	State state;
	Mode mode;
	boolean init;
	public ComponentContainer(Class<?> interfaceClass, Object component) {
		this.component = component;
		this.state = State.IDLE;
		this.init = false;
		this.mode = Mode.CYCLIC;
		this.interfaceClass = interfaceClass;
		
		for (Method m : interfaceClass.getMethods()) {
			
			if (m.isAnnotationPresent(Update.class)) {
				this.event_update.connect(new EventInputSlot(this.component, m));
			}
			if (m.isAnnotationPresent(Init.class)) {
				this.event_init.connect(new EventInputSlot(this.component, m));
			}
			if (m.isAnnotationPresent(Shutdown.class)) {
				this.event_shutdown.connect(new EventInputSlot(this.component, m));
			}
			
		}
	}


	@Override
	protected void finalize() throws Throwable {
		super.finalize();
		this.event_shutdown.emit();
	}
	
	public void setMode(Mode mode) {
		this.mode = mode;
	}
	
	protected void update() {
		if (this.state == State.ACTIVE) {
			this.event_update.emit();
		}
	}
	
	@EventInputPort(name="event_in", event="e_start")
	public void event_input_start() {
		System.out.println("ComponentContainer.event_input_start");
		this.state = State.ACTIVE;
		
		if (this.init == false) {
			this.event_init.emit();
			this.init = true;
		}
	}
	
	@EventInputPort(name="event_in", event="e_stop")
	public void event_input_stop() {
		System.out.println("ComponentContainer.event_input_stop");
		this.state = State.IDLE;
	}
	
	@EventInputPort(name="event_in", event="update")
	public void event_input_update() {
		if (this.mode == Mode.ON_EVENT) {
			this.update();
		}
	}
	
	
	public void update_cycle() {
		if (this.mode == Mode.CYCLIC) {
			this.update();
		}
	}
	
}
